File: week2/STA.py
# -*- coding: utf-8 -*-
"""
Created on Sun Sep  6 13:10:58 2020

@author: Jorge
"""

from __future__ import division

import numpy as np
import matplotlib.pyplot as plt
import pickle

def compute_sta(stim, rho, num_timesteps):
    """Compute the spike-triggered average from a stimulus and spike-train.
    
    Args:
        stim: stimulus time-series
        rho: spike-train time-series
        num_timesteps: how many timesteps to use in STA
        
    Returns:
        spike-triggered average for specified number of timesteps before spike"""
    
    sta = np.zeros((num_timesteps,))

    # This command finds the indices of all of the spikes that occur
    # after 300 ms into the recording.
    spike_times = rho[num_timesteps:].nonzero()[0] + num_timesteps

    # Fill in this value. Note that you should not count spikes that occur
    # before 300 ms into the recording.
    num_spikes = len(spike_times)
    
    # Compute the spike-triggered average of the spikes found.
    # To do this, compute the average of all of the vectors
    # starting 300 ms (exclusive) before a spike and ending at the time of
    # the event (inclusive). Each of these vectors defines a list of
    # samples that is contained within a window of 300 ms before each
    # spike. The average of these vectors should be completed in an
    # element-wise manner.
    # 
    # Your code goes here.
    for i in range(num_spikes):
        AllWindows = stim[(spike_times[i] - num_timesteps):spike_times[i]]
        sta = sta + AllWindows
    sta = sta/num_spikes
    
    return sta

FILENAME = 'c1p8.pickle'
with open(FILENAME, 'rb') as f:
    data = pickle.load(f)
    
stim = data['stim']
rho = data['rho']

sampling_period = 2 # in ms
num_timesteps = 150

sta = compute_sta(stim, rho, num_timesteps)

time = (np.arange(-num_timesteps, 0) + 1) * sampling_period

plt.plot(time, sta)
plt.xlabel('Time (ms)')
plt.ylabel('Stimulus')
plt.title('Spike-Triggered Average')

plt.savefig('spikePython.jpg')

plt.show()

File: week2/STA.m
close all; clear all; clc;

function [ sta ] = compute_sta2( stim, rho, num_timesteps )
%COMPUTE_STA Calculates the spike-triggered average for a neuron that
%            is driven by a stimulus defined in stim. The spike-
%            triggered average is computed over num_timesteps timesteps.
    sta = zeros(num_timesteps, 1);

    % This command finds the indices of all of the spikes that occur
    % after 300 ms into the recording.
    spike_times = find(rho(num_timesteps+1:end)) + num_timesteps;

    % Fill in this value. Note that you should not count spikes that occur
    % before 300 ms into the recording.
    num_spikes = size(spike_times,1);
    
    for i=1:num_spikes
            AllWindows=stim(spike_times(i)-num_timesteps+1:spike_times(i));
            sta = sta + AllWindows;
        end
    
    sta=sta/num_spikes;
end



load('c1p8.mat');

% Fill in these values
sampling_period = 2; % in ms
num_timesteps = 150;

sta = compute_sta2(stim, rho, num_timesteps);

time = -sampling_period*(num_timesteps-1):sampling_period:0; % in ms

figure(1);
plot(time, sta);
xlabel('Time (ms)');
ylabel('Stimulus');
title('Spike-Triggered Average');

saveas(gcf,'spikeMATLAB.png') 